csharp/ql/src/Security Features/CWE-502/UnsafeDeserialization.ql

```diff
@@ -13,7 +13,17 @@
 import csharp
 import semmle.code.csharp.security.dataflow.UnsafeDeserialization::UnsafeDeserialization
 
-from Call deserializeCall, Sink sink
-where deserializeCall.getAnArgument() = sink.asExpr()
+from Call deserializeCall, ObjectMethodSink sink
+where
+  deserializeCall.getAnArgument() = sink.asExpr() and
+  not exists(
+    DataFlow::PathNode constructor, DataFlow::PathNode usage,
+    SafeConstructorTrackingConfig constructorTracking
+  |
+    constructorTracking.hasFlowPath(constructor, usage) and
+    usage.getNode().asExpr().getParent() = sink.asExpr().getParent()
+  )
+  or
+  exists(ConstructorOrStaticMethodSink sink2 | deserializeCall.getAnArgument() = sink2.asExpr())
```

Comment:
It feels strange that this whole query is not symmetric in structure for 
* `ObjectMethodSink` which is defined in the `from` and 
* `ConstructorOrStaticMethodSink` which is in an `exists`. 

I would refactor this along these lines:
```
from Call deserializeCall, Sink sink
where
  deserializeCall.getAnArgument() = sink.asExpr() and
  (
      sink instanceof ObjectMethodSink and
      ...
    or
      sink instanceof ConstructorOrStaticMethodSink
  )
select ...
```